File: hackerrank/problem solving/algorithms/2 - implementation/3 - number line jumps/with morfo library/number_line_jumps.ts
// https://www.hackerrank.com/challenges/kangaroo/problem?isFullScreen=true

import { read } from 'morfo';

function main() {
  const initialState: number[] = read(4, Number);
  console.log(kangaroo(initialState) ? 'YES' : 'NO');
}

// T: O(1)
// S: O(1) extra space
function kangaroo(initialState: number[]): boolean {
  const [x1, v1, x2, v2]: number[] = initialState;

  if (v1 === v2) return x1 === x2;

  const distanceDiff = x2 - x1;
  const velocityDiff = v1 - v2;
  return distanceDiff * velocityDiff >= 0 && distanceDiff % velocityDiff === 0;
}

main();

File: hackerrank/problem solving/algorithms/2 - implementation/3 - number line jumps/with morfo library/number_line_jumps.rb
# https://www.hackerrank.com/challenges/kangaroo/problem?isFullScreen=true

require 'morfo'

def main
  initial_state = Morfo.read(4, &:to_i)
  puts kangaroo?(initial_state) ? "YES" : "NO"
end

# T: O(1)
# S: O(1) extra space
def kangaroo?(initial_state)
  x1, v1, x2, v2 = initial_state

  return x1 == x2 if v1 == v2

  distance_diff = x2 - x1
  velocity_diff = v1 - v2
  distance_diff * velocity_diff >= 0 && (distance_diff % velocity_diff).zero?
end

main if __FILE__ == $0
